File: quantum-safe/PostQuantum.java
package quantum.safe;

import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.Security;
import java.security.Signature;
import org.bouncycastle.pqc.jcajce.provider.BouncyCastlePQCProvider;
import org.bouncycastle.pqc.jcajce.spec.MLDSAParameterSpec;
import org.bouncycastle.pqc.jcajce.spec.MLKEMParameterSpec;

/** Quantum-safe NIST PQC via BouncyCastle: ML-KEM-768 (FIPS 203), ML-DSA-65 (FIPS 204). */
public final class PostQuantum {

    static {
        Security.addProvider(new BouncyCastlePQCProvider());
    }

    /** Generates an ML-KEM-768 key pair (encapsulation uses KEMGenerateSpec/KEMExtractSpec). */
    public static KeyPair mlkem768KeyPair() throws Exception {
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("ML-KEM", "BCPQC");
        kpg.initialize(MLKEMParameterSpec.ml_kem_768);
        return kpg.generateKeyPair();
    }

    public static byte[] mldsa65Sign(byte[] message) throws Exception {
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("ML-DSA", "BCPQC");
        kpg.initialize(MLDSAParameterSpec.ml_dsa_65);
        KeyPair kp = kpg.generateKeyPair();
        Signature sig = Signature.getInstance("ML-DSA", "BCPQC");
        sig.initSign(kp.getPrivate());
        sig.update(message);
        return sig.sign();
    }

    public static void main(String[] args) throws Exception {
        System.out.println("ML-KEM-768 public key: " + mlkem768KeyPair().getPublic().getEncoded().length);
        System.out.println("ML-DSA-65 signature: " + mldsa65Sign("message".getBytes()).length);
    }
}

File: quantum-safe/post_quantum.py
#!/usr/bin/env python3
"""Quantum-safe NIST PQC: ML-KEM-768 (FIPS 203) and ML-DSA-65 (FIPS 204).

Uses liboqs via the `oqs` Python binding (pip install liboqs-python). ML-KEM
(formerly CRYSTALS-Kyber) is key encapsulation; ML-DSA (formerly CRYSTALS-
Dilithium) is digital signatures.
"""
import oqs


def mlkem768_kem() -> bool:
    with oqs.KeyEncapsulation("ML-KEM-768") as kem:
        public_key = kem.generate_keypair()
        ciphertext, shared_secret = kem.encap_secret(public_key)
        return kem.decap_secret(ciphertext) == shared_secret


def mldsa65_sign(message: bytes) -> bool:
    with oqs.Signature("ML-DSA-65") as signer:
        public_key = signer.generate_keypair()
        signature = signer.sign(message)
        return signer.verify(message, signature, public_key)


if __name__ == "__main__":
    print("ML-KEM-768 shared-secret match:", mlkem768_kem())
    print("ML-DSA-65 verified:", mldsa65_sign(b"message"))
